src/test/lib/OwnershipTest.cpp

```diff
@@ -0,0 +1,235 @@
+#include "precomp.h"
+#ifdef QUIC_CLOG
+#include "OwnershipTest.cpp.clog.h"
+#endif
+
+struct OwnershipState {
+    long ShutdownCount{0};
+    CxPlatEvent StateEvent;
+    long StreamAppClose{0};
+    long ConnAppClose{0};
+};
+
+struct ConnectionWrapper {
+    HQUIC Connection {nullptr};
+    ~ConnectionWrapper() noexcept {
+        if (Connection) {
+            MsQuic->ConnectionClose(Connection);
+        }
+    }
+};
+
+struct StreamWrapper {
+    HQUIC Stream {nullptr};
+    ~StreamWrapper() noexcept {
+        if (Stream) {
+            MsQuic->StreamClose(Stream);
+        }
+    }
+};
+
+_Function_class_(QUIC_CONNECTION_CALLBACK)
+static
+QUIC_STATUS
+OwnershipConnCallback(
+    HQUIC,
+    void* Context,
+    QUIC_CONNECTION_EVENT* Event
+    )
+{
+    OwnershipState* State = reinterpret_cast<OwnershipState*>(Context);
+    switch (Event->Type) {
+        case QUIC_CONNECTION_EVENT_CONNECTED:
+            State->StateEvent.Set();
+            break;
+        case QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE:
+            InterlockedIncrement((volatile long*)&State->ShutdownCount);
+            State->StateEvent.Set();
+            break;
+        default:
+            break;
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+void QuicTestRegistrationShutdownBeforeConnOpen()
+{
+    MsQuicRegistration Registration;
+    TEST_TRUE(Registration.IsValid());
+
+    Registration.Shutdown(QUIC_CONNECTION_SHUTDOWN_FLAG_NONE, 0);
+
+    HQUIC Connection = nullptr;
+    QUIC_STATUS Status =
+        MsQuic->ConnectionOpen(
+            Registration,
+            OwnershipConnCallback,
+            nullptr,
+            &Connection);
+
+    TEST_QUIC_STATUS(QUIC_STATUS_OUT_OF_MEMORY, Status);
```

Comment:
I'm not a fan that we get OUT_OF_MEMORY instead of something like INVALID_STATE. I know it would be a (slightly) bigger change, but it's probably worth returning a better error code.

Reply:
I agree. I changed it up, it wasn't too bad.